Web/Scotty/Cookie.hs

```diff
@@ -0,0 +1,364 @@
+    where parse = parseCookiesText . BSL.toStrict . TL.encodeUtf8
+
+
+deleteCookie :: (Monad m, ScottyError e)
+             => Text -- ^ name
+             -> ActionT e m ()
+deleteCookie c = setCookie $ (makeSimpleCookie c "") { setCookieExpires = Just $ posixSecondsToUTCTime 0 }
+
+
+-- | Construct a simple cookie (an UTF-8 string pair with default cookie options)
+makeSimpleCookie :: Text -- ^ name
+                 -> Text -- ^ value
+                 -> SetCookie
+makeSimpleCookie n v = defaultSetCookie { setCookieName  = encodeUtf8 n
+                                        , setCookieValue = encodeUtf8 v
+                                        }
+
+
+-- | Textual cookies. Functions assume UTF8 encoding.
+type CookiesText = [(Text, Text)]
+
+parseCookiesText :: S.ByteString -> CookiesText
+parseCookiesText =
+    map (go *** go) . parseCookies
+  where
+    go = decodeUtf8With lenientDecode
+
+renderCookiesText :: CookiesText -> Builder
+renderCookiesText = renderCookiesBuilder . map (encodeUtf8Builder *** encodeUtf8Builder)
+
+type Cookies = [(S.ByteString, S.ByteString)]
+
+-- | Decode the value of a \"Cookie\" request header into key/value pairs.
+parseCookies :: S.ByteString -> Cookies
+parseCookies s
+  | S.null s = []
+  | otherwise =
+    let (x, y) = breakDiscard 59 s -- semicolon
+     in parseCookie x : parseCookies y
+
+parseCookie :: S.ByteString -> (S.ByteString, S.ByteString)
+parseCookie s =
+    let (key, value) = breakDiscard 61 s -- equals sign
+        key' = S.dropWhile (== 32) key -- space
+     in (key', value)
+
+breakDiscard :: Word8 -> S.ByteString -> (S.ByteString, S.ByteString)
+breakDiscard w s =
+    let (x, y) = S.break (== w) s
+     in (x, S.drop 1 y)
+
+type CookieBuilder = (Builder, Builder)
+
+renderCookiesBuilder :: [CookieBuilder] -> Builder
+renderCookiesBuilder [] = mempty
+renderCookiesBuilder cs =
+    foldr1 go $ map renderCookie cs
+  where
+    go x y = x `mappend` char8 ';' `mappend` y
+
+renderCookie :: CookieBuilder -> Builder
+renderCookie (k, v) = k `mappend` char8 '=' `mappend` v
+
+renderCookies :: Cookies -> Builder
+renderCookies = renderCookiesBuilder . map (byteString *** byteString)
+
+-- | Data type representing the key-value pair to use for a cookie, as well as configuration options for it.
+--
+-- ==== Creating a SetCookie
+--
+-- 'SetCookie' does not export a constructor; instead, use 'defaultSetCookie' and override values (see <http://www.yesodweb.com/book/settings-types> for details):
+--
+-- @
+-- import Web.Cookie
+-- :set -XOverloadedStrings
+-- let cookie = 'defaultSetCookie' { 'setCookieName' = "cookieName", 'setCookieValue' = "cookieValue" }
+-- @
+--
+-- ==== Cookie Configuration
+--
+-- Cookies have several configuration options; a brief summary of each option is given below. For more information, see <http://tools.ietf.org/html/rfc6265#section-4.1.2 RFC 6265> or <https://en.wikipedia.org/wiki/HTTP_cookie#Cookie_attributes Wikipedia>.
+data SetCookie = SetCookie
```

Comment:
What's the reason behind duplicating [the cookie package's definition](https://hackage.haskell.org/package/cookie-0.4.5/docs/Web-Cookie.html#t:SetCookie)?

Reply:
mainly to decouple them, but it's also unlikely that the definition of a web cookie will change in the future, so I will revert to importing 'cookie' rather than vendoring it in.

Reply:
Great, I think that's better